src/main.rs

```diff
@@ -36,7 +36,14 @@ pub mod expansion;
 pub struct Command {
     pub name: &'static str,
     pub help: &'static str,
-    pub main: Box<Fn(&Vec<String>, &mut BTreeMap<String, String>, &mut Vec<Mode>)>,
+    pub main: Box<Fn(&Vec<String>, &mut Shell)>,
+}
+
+/// This struct will contain all of the data structures related to this
+/// instance of the shell.
+pub struct Shell<'a> {
+    pub variables: &'a mut BTreeMap<String, String>,
```

Comment:
I have been using BTreeMap because HashMap is (or at least was not) implemented in libredox. Do you think I should change to HashMap?